### ISSUE TITLE
Add support for TCP prior to airspace boundary

### ISSUE THREAD

**Issue by erikquinn** [openscope/openscope#1856]
Many approach controls have control from the transferring facility for certain actions they can take _prior_ to the aircraft actually entering their airspace. KPHL is a good example-- just looking at the JIIMS side, Washington Center grants Philadelphia TRACON control for:
- descent within 10nm of VCN/JIIMS
- 40 degree right turns on or north of the VCN115 radial
- 40 degree left turns on or north of the SIE.JIIMS3 transition when north of V268

While complex implementations like this may not be practical in their entirety, it is equally impractical to expect the control position to function effectively while waiting until the aircraft fully enters the airspace. Simply allowing the aircraft to become fully controllable within a set distance of a defined point should be sufficient to cover the majority of use cases (while obviously, not all). In the case of the JIIMS arrivals for example, giving PHL full control within 10nm of JIIMS would make things more workable.

**Comment by CanisLupus518** [comment]
> As an alternative to #1856, I would also be willing to accommodate a "F this, give me the planes" button-- something like "All arrivals are my control", where it defaults to "off" and you can turn it "on", to allow manipulation of aircraft who have not reached the airspace. This is a lesser solution obviously, but is a shorter path we'd be more willing to accept than false manipulation of airspace.

@erikquinn you mentioned this in #1854 but I thought this would be a better place to reply. What about a slight compromise. Howabout a "Control Range" button? Instead of just saying "F this, give me the planes" how about a more nuanced slider that allows a configurable range outside the airspace to become controllable? For instance, I might say I want handoffs 10 nm from my airspace laterally or within 1500 ft vertically. This would be easier to implement if we just looked at the airspace as a whole rather than per-STAR or per `SpawnPattern` rules.  Thoughts?

**Comment by erikquinn** [comment]
That's a little too in the middle for me. There's exactly two options on how to support it:

1. Specify all specific places where we have control early. In reality, we should not allow any turns until reaching the airspace, except where specifically told otherwise. This is the highest fidelity option, and can be done with varying degrees of complexity. I would advocate for the simplest one of specifying a fix or lat/lon and a distance within which we have control (this should cover well over 90% of use cases sufficiently, including all of PHL's).

2. To save effort in developing an implementation to option 1, add an option that allows the user to override the "he's not in your airspace yet" logic. This can be constrained programmatically as needed, but we should not give the user (or airport designer) the autonomy to decide "how far out" they have control. We accept that it's a non-realistic workaround that solves the problem the fastest. To make the range user-configurable would go against the "this isn't real-world accurate, but it is the provided workaround for the lack of option 1 existing".

Basically, either let them control everything, or do a proper modeling of early transfers of control-- but nothing in the middle.

**Comment by CanisLupus518** [comment]
@erikquinn What if we added a new "type" of airspace in the section of the airport config that defines the airspace boundaries? This way rather than specifying a range from a fix, the author of an airport file can provide the exact shape of "handoff" space where control is allowed (useful if such a boundary isn't defined by a fix, or for handling `SpawnPatterns` that aren't on a STAR .  This space could also be displayed explicitly as maybe a different color in the airspace display. This way there really wouldn't be any new logic needed as an aircraft would become controllable once it enters that airspace. Some new logic might be needed if you want it to only effect arrivals , or if you want to be able to define control limits associated with that space.

**Comment by erikquinn** [comment]
I suppose that might work! Does seem slightly more difficult for the typical airport designer to update and maintain. Since usually it's from an SOP/LOA that you'll find the information, where it's written like "turns north of V whatever", "descent north of HPW", etc, making it easier to describe it textually than by means of a polygon of GPS coordinates.

Regardless, you are definitely correct that it wouldn't be a fully exact and accurate representation in all cases, where a polygon could be. I would just counter that it will almost always be able to yield a more-or-less indistinguishably accurate area, and makes implementing it much much easier for the airport people.

One other thing that comes to mind is that you don't always have control of everything in that area. Sometimes only of arrivals to one airport but not the other, when they're on the same route and altitude. (Admittedly we don't allow that in airports now because openscope would crash them, but just thinking future wise)

Certainly not closed minded about your way, but I am definitely hesitant. For some reason my gut tells me it could get messy, perhaps because of parts of the code I don't even remember. And given that I'm not doing the work, I'll accept either, while at this moment I do think the wisest play is accepting an array of fix+distance pairs for each spawn pattern.

**Comment by CanisLupus518** [comment]
The only thing I get hung up on here with your suggestion is that the `SpawnPattern` then becomes integral to checking for controllability of individual aircraft, which to me seems like it might be out of scope, unless we are ok with passing that information along to the aircraft itself at spawn time for a check on it later.

**Comment by erikquinn** [comment]
@CanisLupus518 Passing that sort of condition to the `AircraftModel` is totally fine, as that is already where the controllableness (😂) of the aircraft is stored, checked, and updated.
```
AircraftModel.update() -> ._updateAircraftVisibility() -> ._updateControllableStatus() -> sets .isControllable

Note: ._updateAircraftVisibliity() calls
    AircraftModel.isInsideAirspace(AirportController.airport_get())
    to determine when "this" is controllable.
```
